server/src/main/java/io/crate/execution/support/RetryRunnable.java

```diff
@@ -21,31 +21,36 @@
 
 package io.crate.execution.support;
 
-import io.crate.common.unit.TimeValue;
+import java.util.Iterator;
+import java.util.concurrent.Executor;
+import java.util.concurrent.TimeUnit;
+
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
 import org.elasticsearch.common.util.concurrent.EsRejectedExecutionException;
+import org.elasticsearch.threadpool.Scheduler;
+import org.elasticsearch.threadpool.ThreadPool;
 
-import java.util.Iterator;
-import java.util.concurrent.Executor;
-import java.util.concurrent.ScheduledExecutorService;
-import java.util.concurrent.TimeUnit;
+import io.crate.common.unit.TimeValue;
 
-public class RetryRunnable implements Runnable {
+public class RetryRunnable implements Runnable, Scheduler.Cancellable {
 
     private static final Logger LOGGER = LogManager.getLogger(RetryRunnable.class);
 
+    private final ThreadPool threadPool;
+    private final String executorName;
     private final Executor executor;
-    private final ScheduledExecutorService scheduler;
     private final Iterator<TimeValue> delay;
     private final Runnable retryCommand;
+    private Scheduler.Cancellable cancellable;
```

Comment:
I think this might need to be volatile. Due to the re-schedule `cancellable` is set from a different threads than the read access happening in `cancel`

Reply:
good catch, thx. will create a follow up fix.

Reply:
addressed with https://github.com/crate/crate/pull/12481